File: Assets/_Scripts/Commons/StateUtils/StateMachine.cs
namespace RageCure.StateUtils
{
    public class StateMachine<T>
    {
        public State<T> CurrentState { get; set; }
        public State<T> PreviousState { get; set; }
        public State<T> GlobalState { get; set; }
        public StateMachine(State<T> currentState)
        {
            CurrentState = currentState;
        }

        public void Update(T entity)
        {
            if (GlobalState != null)
                GlobalState.Execute(entity);
            if (CurrentState != null)
                CurrentState.Execute(entity);
        }
        public void FixedUpdate(T entity)
        {
            if (GlobalState != null)
                GlobalState.FixedExecute(entity);
            if (CurrentState != null)
                CurrentState.FixedExecute(entity);
        }

        public void ChangeState(T entity, State<T> newState)
        {
            if (newState == null) return;
            PreviousState = CurrentState;
            CurrentState.Exit(entity);
            CurrentState = newState;
            CurrentState.Enter(entity);
        }
        public void RevertToPreviousState(T entity)
        {
            ChangeState(entity, PreviousState);
        }
        public bool IsInState(State<T> state)
        {
            return state == CurrentState;
        }
    }

}

File: Assembly-FSharp/Assembly-FSharp/Commons/StateMachine.fs
namespace RageCure.StateUtils

open System.Collections.Concurrent

type StateMachine<'a> (currentState: IState<'a>) =

    let mutable currentState: IState<'a> = currentState
    let mutable previousState: IState<'a> option = None
    let mutable globalState: IState<'a> option = None

    member this.CurrentState
        with get() = currentState

    member this.Update(entity: 'a) =
        match globalState with
        | Some gs -> gs.Execute entity; currentState.Execute entity
        | None -> currentState.Execute entity
        
    member this.FixedUpdate(entity: 'a) =
        match globalState with
        | Some gs -> gs.FixedExecute entity; currentState.Execute entity
        | None -> currentState.FixedExecute entity

    member this.ChangeState (newState: IState<'a>) (entity: 'a) =
        previousState <- Some currentState
        currentState.Exit entity
        currentState <- newState
        currentState.Enter entity

    member this.RevertToPreviousState (entity: 'a) =
        match previousState with
        | None -> ()
        | Some ps -> (this.ChangeState ps entity)

    member this.IsInState (state: IState<'a>) =
        state = currentState
    
